### PR TITLE
Mix in controller properties for `Piezo`

### PR DESCRIPTION
**Pull request by bcjordan**
Allow for custom piezo controller following [led controller](https://github.com/rwaldron/johnny-five/blob/master/lib/led/led.js#L142) property mix-in pattern. Mostly to support overriding of `setTone`/`noTone` per https://github.com/rwaldron/playground-io/pull/1, but may be of interest for #815.

@rwaldron let me know if you'd prefer I stub out more or use a different approach for mixing in, could do any of:

1. Stub out default controller behavior (add a `Controllers` map with an empty `DEFAULT`)
1. Move default piezo tone/no-tone behavior into the `DEFAULT` controller
1. Use some sort of different controller overriding pattern

TODO:

- [x] 1, 2, 3 above
- [x] Write test covering controller behavior

### CODE REVIEW

**Comment by rwaldron** [comment]
I'd like to see the complete migration, if you're up for doing it :)

**Comment by rwaldron** [comment]
This will get you started: https://github.com/rwaldron/johnny-five/tree/piezo-controller I've left all the nitty-gritty for you to do :)

Here's how to test: `grunt nodeunit:just:piezo && grunt nodeunit:just:extended/piezo`

**Comment by bcjordan** [comment]
Perfect, on this tomorrow morning, thanks!

File: lib/piezo.js [hunk 1]
```diff
@@ -355,4 +423,9 @@ Piezo.prototype.stop = function() {
   return this;
 };
 
+if (IS_TEST_MODE) {
+  Piezo.frequencyToTone = frequencyToTone;
```
Comment (bcjordan) [inline lib/piezo.js]: Any preferred way to expose helper util methods for testing?

File: lib/piezo.js [hunk 2]
```diff
@@ -229,38 +301,39 @@ Piezo.prototype.note = function(note, duration) {
...
-    value = value === 1 ? 0 : 1;
-    this.io.digitalWrite(this.pin, value);
+  return this.frequency(toneToFrequency(tone), duration);
+};
 
-    if ((timer.difTime / 1000000) > duration) {
-      clearTimer.call(this);
-    }
-  }.bind(this), null, tone + "u", function() {});
+Piezo.prototype.frequency = function(frequency, duration) {
```
Comment (bcjordan) [inline lib/piezo.js]: One quirky part of this implementation I'd like thoughts on / alternatives, this requires the given controller to define at least one of `frequency` and `tone`. The Circuit Playground's sysex API, for example, works with `Piezo.prototype.frequency`-equivalent values (called `tone` in [Arduino land](https://www.arduino.cc/en/Reference/Tone)), while the internals of the default/digital piezo implementation work at the `Piezo.prototype.tone` level and treat frequency as the helper.

The intent with this approach is to reconcile controller implementations wanting to work with either tone OR frequency, without implementations needing to re-define or import the conversion methods.

Alternatively we could treat `Piezo.prototype.tone` as the only controller override-able method, just not sure whether that's more likely to be cumbersome for future controllers than `Piezo.prototype.frequency`.

**Comment by bcjordan** [comment]
@rwaldron alright, ready for another look! & added a couple of line note comments with design choice Qs.

File: lib/piezo.js (hunk 1 above)
Comment (rwaldron) [inline lib/piezo.js]: You can use this same silliness if you want. It's "meh", but works and stays out of the way of non-test-runtime API

**Comment by rwaldron** [comment]
@bcjordan Landed!

**Comment by bcjordan** [comment]
@rwaldron awesome, thanks for your help on this!

**Pull request merged**